File: php/02-open-closed-principle.php
<?php

// Open Closed Principle Violation

class Programmer
{
    public function code()
    {
        return 'coding';
    }
}

class Tester
{
    public function test()
    {
        return 'testing';
    }
}

class ProjectManagement
{
    public function process($member)
    {
        if ($member instanceof Programmer) {
            $member->code();
        } elseif ($member instanceof Tester) {
            $member->test();
        };

        throw new Exception('Invalid input member');
    }
}

// Refactored

interface IWorkable
{
    public function work();
}

class Programmer implements IWorkable
{
    public function work()
    {
        return 'coding';
    }
}

class Tester implements IWorkable
{
    public function work()
    {
        return 'testing';
    }
}

class ProjectManagement
{
    public function process(IWorkable $member)
    {
        return $member->work();
    }
}

File: c#/02-OpenClosedPrinciple.cs

// Open Closed Principle Violation

class Programmer
{
    public string Code()
    {
        return "Coding";
    }
}

class Tester
{
    public string Test()
    {
        return "testing";
    }
}

class ProjectManagement
{
    public string Process(object member)
    {
        if (typeof(member) == Programmer) {

            return (member as Programmer).Code();

        } else if (typeof(member) == Tester) {

            return (member as Tester).Test();
        };

        throw new System.Exception("Invalid input member");
    }
}

// Refactored

interface IWorkable
{
    Work();
}

class Programmer : IWorkable
{
    public Work()
    {
        return "Coding";
    }
}

class Tester : IWorkable
{
    public Work()
    {
        return "Testing";
    }
}

class ProjectManagement
{
    public string Process(IWorkable member)
    {
        return member.Work();
    }
}
